docs: Implement style guide rules - active voice, misc #2 (#23273)
MystenLabs/sui: docs/content/concepts/sui-architecture/protocol-upgrades.mdx

---
title: Protocol Upgrades
description: The Sui protocol, framework, and execution engine are frequently extended to include new functionality and bug fixes. The upgrade process ensures all clients use the same source.
---

The Sui protocol, framework, and execution engine are frequently extended to include new functionality and bug fixes. This functionality is added as new code released to validator operators as part of regular software releases. The Sui protocol requires that all validators agree on the results of executing each transaction.  

> This creates a challenge: how do we release code that changes transaction execution, given that operators cannot all upgrade at the same instant? Further, how do we ensure that all transaction history can be replayed even after functionality has changed?  

Author: ```suggestion
This creates a challenge: how is code released that changes transaction execution, given that operators cannot all upgrade at the same instant? Further, how is it ensured that all transaction history can be replayed even after functionality has changed?  
```

To solve this, Sui uses a process called protocol upgrades.

## Protocol upgrade process

The process for protocol upgrades includes these steps:

> 1. A developer codes the new feature but restricts access to it with a feature flag, a boolean configuration variable initially set to false.  

Author: Yea agreed now that you bring that up, I fixed the other PR before I merged, will fix this one as well
